src/cis2/mod.rs

```diff
@@ -0,0 +1,352 @@
+        };
+        let tx = transactions::send::make_and_sign_transaction(
+            signer,
+            transaction_metadata.sender_address,
+            transaction_metadata.nonce,
+            transaction_metadata.expiry,
+            transaction_metadata.energy,
+            payload,
+        );
+        let bi = transactions::BlockItem::AccountTransaction(tx);
+        let hash = self.client.send_block_item(&bi).await?;
+        Ok(hash)
+    }
+
+    /// Send a CIS2 updateOperator transaction.
+    /// Construct and send a CIS2 updateOperator smart contract update
+    /// transaction given a list of CIS2 UpdateOperators. Returns a Result
+    /// with the transaction hash.
+    ///
+    /// # Arguments
+    ///
+    /// * `signer` - The account keys to use for signing the smart contract
+    ///   update transaction.
+    /// * `sender_address` - The account address of the signing keys.
+    /// * `transaction_metadata` - Metadata for constructing the transaction.
+    /// * `updates` - A list of CIS2 UpdateOperators to update.
+    #[allow(clippy::too_many_arguments)]
+    pub async fn update_operator(
+        &mut self,
+        signer: &impl transactions::ExactSizeTransactionSigner,
+        transaction_metadata: Cis2TransactionMetadata,
+        updates: Vec<UpdateOperator>,
+    ) -> anyhow::Result<sdk_types::hashes::TransactionHash, Cis2TransactionError> {
+        let parameter = UpdateOperatorParams::from(updates);
+        let bytes = concordium_contracts_common::to_bytes(&parameter);
+        let contract_name = self.contract_name.contract_name();
+        let receive_name =
+            smart_contracts::ReceiveName::try_from(format!("{}.updateOperator", contract_name))?;
+
+        let payload = transactions::Payload::Update {
+            payload: transactions::UpdateContractPayload {
+                amount: transaction_metadata.amount,
+                address: self.address,
+                receive_name,
+                message: smart_contracts::Parameter::from(bytes),
+            },
+        };
+        let tx = transactions::send::make_and_sign_transaction(
+            signer,
+            transaction_metadata.sender_address,
+            transaction_metadata.nonce,
+            transaction_metadata.expiry,
+            transaction_metadata.energy,
+            payload,
+        );
+        let bi = transactions::BlockItem::AccountTransaction(tx);
+        let hash = self.client.send_block_item(&bi).await?;
+        Ok(hash)
+    }
+
+    /// Invoke the CIS2 balanceOf query given a list of BalanceOfQuery.
+    ///
+    /// Note: the query is executed locally by the node and does not produce a
+    /// transaction on-chain.
+    ///
+    /// # Arguments
+    ///
+    /// * `block_hash` - The hash of the block to locally execute the query
+    ///   after.
+    /// * `queries` - A list queries to execute.
+    pub async fn balance_of(
+        &mut self,
+        block_hash: &sdk_types::hashes::BlockHash,
```

Comment:
By making it an option?
I think it is best for the user to be eksplicit about which block they care about, since that would depend on the usecase.
It might be better to add functions for getting the `best_block` and `last_finalized_block` directly.

Reply:
Yes, or have a general adapter `with_last_final` although the type signature of that one is something to behold.